twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt

```diff
@@ -126,20 +139,86 @@ class TwingraphServiceImpl(
         }
       }
     }
-
-    val redisGraphId = "${graphId}:${twinGraphQuery.version}"
-
+    val redisGraphId = redisGraphKey(graphId, twinGraphQuery)
     csmJedisPool.resource.use { jedis ->
       val redisGraphMatchingKeys = jedis.keys(redisGraphId)
       if (redisGraphMatchingKeys.size == 0) {
         throw CsmResourceNotFoundException("No graph found with id: $redisGraphId")
       }
     }
+  }
+
+  override fun query(
+      organizationId: String,
+      graphId: String,
+      twinGraphQuery: TwinGraphQuery
+  ): String {
+    checkTwinGraphPrerequisites(organizationId, graphId, twinGraphQuery)
     val redisGraph = RedisGraph(csmJedisPool)
     val resultSet =
         redisGraph.query(
-            redisGraphId, twinGraphQuery.query, csmPlatformProperties.twincache.queryTimeout)
-
+            redisGraphKey(graphId, twinGraphQuery),
+            twinGraphQuery.query,
+            csmPlatformProperties.twincache.queryTimeout)
     return resultSet.toJsonString()
   }
+
+  override fun bulkQuery(
+      organizationId: String,
+      graphId: String,
+      twinGraphQuery: TwinGraphQuery
+  ): TwinGraphHash {
+    checkTwinGraphPrerequisites(organizationId, graphId, twinGraphQuery)
+    val redisGraphKey = redisGraphKey(graphId, twinGraphQuery)
+    var bulkQueryKey = bulkQueryKey(graphId, twinGraphQuery)
+    GlobalScope.launch(SecurityCoroutineContext()) {
+      csmJedisPool.resource.use { jedis ->
+        val keyExists = jedis.exists(bulkQueryKey.first)
+        if (keyExists) {
+          return@launch
+        }
+        val redisGraph = RedisGraph(csmJedisPool)
+        val resultSet = redisGraph.query(redisGraphKey, twinGraphQuery.query)
+        val zip =
+            zipBytesWithFileNames(
+                mapOf("bulkQuery.json" to resultSet.toJsonString().toByteArray(UTF_8)))
+        jedis.setex(bulkQueryKey.first, csmPlatformProperties.twincache.queryBulkTTL, zip)
+      }
+    }
+    return TwinGraphHash(bulkQueryKey.second)
+  }
+
+  override fun downloadGraph(organizationId: String, graphQueryHash: String): Resource {
+    val organization = organizationService.findOrganizationById(organizationId)
+    csmRbac.verify(organization.getRbac(), PERMISSION_READ)
+
+    var bulkQueryId = bulkQueryKey(graphQueryHash)
+    csmJedisPool.resource.use { jedis ->
+      if (!jedis.exists(bulkQueryId)) {
+        throw CsmResourceNotFoundException("No graph found with hash: $graphQueryHash  Try later")
+      } else if (jedis.ttl(bulkQueryId) < 0) {
+        throw CsmResourceNotFoundException(
+            "Graph with hash: $graphQueryHash is expired. Try to repeat bulk query")
+      }
+    }
+    val httpServletResponse =
+        ((RequestContextHolder.getRequestAttributes() as ServletRequestAttributes).response)
+    val contentDisposition = ContentDisposition.builder("attachment").filename("Graph.zip").build()
```

Comment:
Could be interesting to have a specific zip file name (regarding query maybe) instead of a fixed value ;)